### DISCUSSION TITLE
Change emissive factor in js

### DISCUSSION THREAD

**Discussion by Thyronia** [google/model-viewer#3050] [Q&A] [answered]
Hello everyone !

I've been trying to change the emissive factor in my script but I can't find a way to do it...

The code so far :

```
var emissive = [255,0,0];	
modelViewer.model.materials[2].emissiveFactor = emissive;
```

modelViewer.model.materials[2] refers to the correct texture (I can change the base color without problem for example).

Does anyone know how to change this? 

Thanks a lot !

**Comment by slightHYP** [comment, marked as answer]
I used the setter function to apply the new value.

In your case it should be something like this:
`modelViewer.model.materials[2].setEmissiveFactor(emissive);`

**Reply by Thyronia** [reply]
Thank you so much ! I couldn't find this kind of info in the docs, can you tell me where you found this?

**Reply by slightHYP** [reply]
I had your same problem some months ago, so I had a look at the source code. In particular inside this script:

https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/features/scene-graph/api.ts

here I found what is currently exposed by model-viewer.

@elalish Is this the correct script that provides an overview of all the public available features, methods and attributes? Is it correct, right?

**Reply by Thyronia** [reply]
Great, I see it ! Thank you again :)

**Reply by elalish** [reply]
Yes, that's right. We need to update the public docs.

**Comment by gabrielgaldinomagalhaes** [comment]
Does the emissive lights the other pieces of the object? I mean, if I have the model-viewer exposure set up to 0, and a glb with a black box and a sphere inside of the box, if my sphere has an emissive material, should the sphere light the box ?

**Reply by elalish** [reply]
No, that's too expensive to ask a rasterizer to do. For that you'd need a path-tracer and a much lower frame rate. You can fake it by baking an emissive texture on the receiving objects.

**Reply by gabrielgaldinomagalhaes** [reply]
Alright! I got it! I used an emissive texture inside of the material, and it work kind the way. But now I want to change the emissiveIntensity of these emissive texture of this specific material.

I was testing some approaches and identified that, on console, when I use:
modelViewer.model.materials

If I reach the material index 1. I can reach the emissiveTexture inside of this material (using model.materials[1].emissiveTexture, there is no error in the console).
But once I tried to reach something inside the emissiveTexture that is inside the material[1], it gives me "undefined".

In the console I can see that inside the emissiveTexture there is a set of materials, and inside of this set I can see the emissiveIntensity.
 
If I change this value using bruteforce straight in the console it works, in the way I wanted it to work (see the screenshot bellow).
Is there anyway to reach the emissiveIntensity that is inside the materials that are inside the emissiveTexture? I don't know if you get it.

But check the console prints that I sent bellow.
In the printscreens you can see me reaching the value I want, on the console, but I can't do that using codes. Because once I try to get something inside the  model.materials[1].emissiveTexture, it gives me "undefined".

![image](https://user-images.githubusercontent.com/45290323/222651117-02a935b4-15e9-4222-9d76-51dc759f72fc.png)

In this image you can see the emissiveIntensity as 1, and the result in the object. Dark!
![image](https://user-images.githubusercontent.com/45290323/222652347-92d111aa-74ba-45e7-a049-2884ddb1495f.png)

In this image you can the emissiveIntensity as 100, and the result in the object. A lot lighter!
![image](https://user-images.githubusercontent.com/45290323/222652428-b3d3ce98-0542-4335-80ff-531d25892d42.png)

**Reply by gabrielgaldinomagalhaes** [reply]
According to what I see in the structure in the console, I could reach that using something like this
modelViewer.model.materials[1].emissiveTexture.materials[0].emissiveIntensity

It work till here, modelViewer.model.materials[1].emissiveTexture
But once I try to get something inside the emissiveTexture, it gives me undefined!

**Reply by elalish** [reply]
Any time you see a `Symbol`, you've left our public API - please refer to the docs: https://modelviewer.dev/examples/scenegraph/#pickMaterialExample - scroll down and you'll see this is accessed via `modelViewer.model.materials[1].setEmissiveFactor(rgb);`

**Reply by gabrielgaldinomagalhaes** [reply]
Thank you for your reply, yes, I saw the setEmissiveFactor(), I tested it already! But it allows me to change only the RGB and not the intensity itself. I reckon it would be a 4th parameter in the function. And was exactly after the setEmissiveFactor() that I thought changing the intensity would also be possible. So it is not, not straight on model-viewer, right ?

**Reply by elalish** [reply]
Did you try values > 1?
